### PR TITLE
feat: BIP39Wallet

### PR DESCRIPTION
**Pull request by xiboXD**

### CODE REVIEW

File: src/BIP39Wallet.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        public static BlockchainWallet GetWalletByPrivateKey(string privateKey)
+        {
+            var keybyte = StringToByteArray(privateKey);
+            Array.Resize(ref keybyte, 32);
+            var key = new Key(keybyte, -1, false);
+            var publicKey = key.PubKey;
+            var address =  Address.FromPublicKey(publicKey.ToBytes()).ToString().Trim('\"');
+            return new BlockchainWallet(address, privateKey, null!, publicKey.ToHex());
+        }
+        public static byte[] Sign(byte[] privateKey, byte[] hash)
```
Comment (argfoo-web3) [inline src/BIP39Wallet.cs:58]: I was thinking, if we want to make this repo generic, we might need to make these methods virtual for people to extend from and implement their signing function. If we do it this way, it is basically for the AElf ecosystem.

File: test/BIP39Wallet.Tests.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            byte[] byteArray = new byte[length / 2];
+
+            for (int i = 0; i < length; i += 2)
+            {
+                byteArray[i / 2] = Convert.ToByte(hexString.Substring(i, 2), 16);
+            }
+
+            return byteArray;
+        }
+        public static string ToHexString(byte[] bytes)
```
Comment (argfoo-web3) [inline test/BIP39Wallet.Tests.cs]: Duplicated code here. Probably need to include the Helper class assembly to use it.

Comment (xiboXD) [reply test/BIP39Wallet.Tests.cs]: deleted

File: src/BIP39Wallet.cs (hunk 1 above)
Comment (xiboXD) [reply src/BIP39Wallet.cs:58]: done, changed to virtual

Comment (gldeng) [reply src/BIP39Wallet.cs:58]: Why would someone use the library and override the methods?

File: src/BIP39Wallet.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+using System.Text;
+using AElf.Types;
+using System.Security.Cryptography;
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+
+namespace BIP39Wallet
+{
+    [SuppressMessage("ReSharper", "UnusedAutoPropertyAccessor.Global")]
```
Comment (gldeng) [inline src/BIP39Wallet.cs]: This is not needed.

File: src/BIP39Wallet.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+            public BlockchainWallet(string address, string privateKey, string mnemonic, string publicKey)
+            {
+                Address = address;
+                PrivateKey = privateKey;
+                Mnemonic = mnemonic;
+                PublicKey = publicKey;
+            }
+        }
+        public virtual string ConvertMnemonicToSeedHex(Mnemonic mnemonic, string password)
```
Comment (gldeng) [inline src/BIP39Wallet.cs]: Don't make it virtual.

File: src/BIP39Wallet.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            var key = rfc2898DerivedBytes.GetBytes(64);
+            var hex = BitConverter
+                .ToString(key)
+                .Replace("-", "")
+                .ToLower();
+
+            return hex;
+        }
+
+        public virtual BlockchainWallet CreateWallet(int strength, Language language, string password)
```
Comment (gldeng) [inline src/BIP39Wallet.cs]: All parameters are not used.

File: src/BIP39Wallet.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            var mnemonic = new Mnemonic(Wordlist.English, WordCount.Twelve);
+
+            var seed = ConvertMnemonicToSeedHex(mnemonic, "");
+            var masterKeyPath = new KeyPath("m/44'/1616'");
```
Comment (gldeng) [inline src/BIP39Wallet.cs]: The string can be extracted out and made as constant and give it a proper name.

File: src/BIP39Wallet.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            var privateKey = wallet.PrivateKey;
+            var newKey = new Key(privateKey.ToBytes(), -1, false);
+            var publicKey = newKey.PubKey;
+
+            // Act
+            var address = Address.FromPublicKey(publicKey.ToBytes()).ToString().Trim('\"');
+            return new BlockchainWallet(address, privateKey.ToHex(), mnemonic, publicKey.ToString());
+        }
+        
+        static byte[] StringToByteArray(string hexString)
```
Comment (gldeng) [inline src/BIP39Wallet.cs]: We have a `Helper` class. Why don't we move it there?

File: src/BIP39Wallet.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            int length = hexString.Length;
```
Comment (gldeng) [inline src/BIP39Wallet.cs]: Use `var`.

File: src/Helper.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+using System.Text;
+
+namespace BIP39Wallet
+{
+    public static class Helper
```
Comment (gldeng) [inline src/Helper.cs]: If we don't intend to expose this class, make it internal.

File: test.py [hunk 10]
```diff
@@ -0,0 +1,2 @@
+s = '045f386280af0f51a54ec4afdea2eb49ccf73bb0459d088f89ec0ae759fc7cbebb5b712139a43ef62ea10b4c0f737ea104d8d0a0c8d010561a7e8db7c71292a286'
+print(len(s))
```
Comment (gldeng) [inline test.py]: What's this for?

**Review by gldeng — COMMENTED** [review]
We need to create documentation for each public method.

**Review by gldeng — COMMENTED** [review]
All methods can all be static. So make the class static.

**Review by gldeng — COMMENTED** [review]
Let's move `BlockchainWallet` out. Don't make it a nested class. And the name is bad, let's pick a better name.

File: src/BIP39Wallet.cs (hunk 1 above)
Comment (argfoo-web3) [reply src/BIP39Wallet.cs:58]: Was thinking they could override the signing format if they are using the library for other chains. But I guess in that case, then they should import the chain's relevant library.

File: src/BIP39Wallet.cs (hunk 3 above)
Comment (xiboXD) [reply src/BIP39Wallet.cs]: deleted

File: src/BIP39Wallet.cs (hunk 5 above)
Comment (xiboXD) [reply src/BIP39Wallet.cs]: modified this part

File: src/BIP39Wallet.cs (hunk 4 above)
Comment (xiboXD) [reply src/BIP39Wallet.cs]: chaged to static

File: src/BIP39Wallet.cs (hunk 6 above)
Comment (xiboXD) [reply src/BIP39Wallet.cs]: extracted out in BIP39WalletCOnstants.cs, name it as AElfPath

File: src/BIP39Wallet.cs (hunk 7 above)
Comment (xiboXD) [reply src/BIP39Wallet.cs]: moved

File: src/BIP39Wallet.cs (hunk 8 above)
Comment (xiboXD) [reply src/BIP39Wallet.cs]: changed

File: src/Helper.cs (hunk 9 above)
Comment (xiboXD) [reply src/Helper.cs]: made it internal

File: test.py (hunk 10 above)
Comment (xiboXD) [reply test.py]: redundant file, deleted

**Comment by xiboXD** [comment]
> Let's move `BlockchainWallet` out. Don't make it a nested class. And the name is bad, let's pick a better name.

moved out, and named it as Account

> All methods can all be static. So make the class static.

all changed

**Pull request closed without merging**